File: docs/reference/modules/ROOT/examples/test/xtdb/docs/examples/configuration/configuration_test.clj
(ns xtdb.docs.examples.configuration.configuration-test
  (:require [clojure.test :as t]
            [xtdb.fixtures :as fix :refer [*api*]]))

;; tag::import[]
(require '[xtdb.api :as xt]
         '[clojure.java.io :as io])
;; end::import[]

(t/deftest explicit
  (with-open [node
              ;; tag::from-explicit[]
              (xt/start-node {
                              ;; Configuration Map
                              })
              ;; end::from-explicit[]
              ]
    (t/is true)))

(t/deftest from-file
  (with-open [node
              ;; tag::from-file[]
              (xt/start-node (io/file "resources/config.json"))
              ;; end::from-file[]
              ]
    (t/is true)))

(t/deftest from-resource
  (with-open [node
              ;; tag::from-resource[]
              (xt/start-node (io/resource "config.json"))
              ;; end::from-resource[]
              ]
    (t/is true)))

(t/deftest http-server
  (with-open [node
              ;; tag::http-server[]
              (xt/start-node {:xtdb.http-server/server {:port 3000}})
              ;; end::http-server[]
              ]
    (t/is true)))

(comment "Not testing this one as it requires real info!"
         (t/deftest override-module
           (with-open [node
                       ;; tag::override-module[]
                       (xt/start-node {:xtdb/document-store {:xtdb/module 'xtdb.s3/->document-store
                                                           :bucket "my-bucket"
                                                           :prefix "my-prefix"}})
                       ;; end::override-module[]
                       ]
             (t/is true))))

(t/deftest nested-modules
  (with-open [node
              ;; tag::nested-modules-0[]
              (xt/start-node {:xtdb/tx-log {:kv-store {:xtdb/module 'xtdb.rocksdb/->kv-store
                                                     :db-dir (io/file "/tmp/txs")}}
                              ;; end::nested-modules-0[]
                              })]
    (comment [( ("This obviously won't run so putting in a comment"
                 ;; tag::nested-modules-1[]
                 :xtdb/document-store { }
                 :xtdb/index-store { }
                 ;; end::nested-modules-1[]
                 ) {
                    ;; tag::nested-modules-2[]
                    })
              ;; end::nested-modules-2[]
              ])
    (t/is true)))

(t/deftest sharing-modules
  (with-open [node
              ;; tag::sharing-modules[]
              (xt/start-node {:my-rocksdb {:xtdb/module 'xtdb.rocksdb/->kv-store
                                           :db-dir (io/file "/tmp/rocksdb")}
                              :xtdb/tx-log {:kv-store :my-rocksdb}
                              :xtdb/document-store {:kv-store :my-rocksdb}})
              ;; end::sharing-modules[]
              ]
    (t/is true)))

File: docs/reference/modules/ROOT/examples/test/xtdb/docs/examples/configuration/ConfigurationTest.java
package xtdb.docs.examples.configuration;

import org.junit.*;
import static org.junit.Assert.*;
import java.io.IOException;
import java.io.File;
import java.net.URL;
import java.net.URISyntaxException;

// tag::import[]
import xtdb.api.IXtdb;
// end::import[]

public class ConfigurationTest {

    @Test
    public void fromFile() {
        // tag::from-file[]
        IXtdb xtdbNode = IXtdb.startNode(new File("resources/config.json"));
        // end::from-file[]

        close(xtdbNode);
    }

    @Test
    public void fromResource() {
        try {
            URL url = MyApp.class.getResource("config.json");
            File file = new File(url.toURI());
            assertTrue(file.exists());
        }
        catch (URISyntaxException e) {
            fail();
        }
        // tag::from-resource[]
        IXtdb xtdbNode = IXtdb.startNode(MyApp.class.getResource("config.json"));
        // end::from-resource[]

        close(xtdbNode);
    }

    @Test
    public void withConfigurator() {
        // tag::from-configurator[]
        IXtdb xtdbNode = IXtdb.startNode(n -> {
            // ...
        });
        // end::from-configurator[]

        close(xtdbNode);
    }

    @Test
    public void httpServer() {
        // tag::http-server[]
        IXtdb xtdbNode = IXtdb.startNode(n -> {
            n.with("xtdb.http-server/server", http -> {
                http.set("port", 3000);
            });
        });
        // end::http-server[]

        close(xtdbNode);
    }

    // Not testing this one as it requires real information
    public void overrideModuleImplementation() {
        // tag::override-module[]
        IXtdb xtdbNode = IXtdb.startNode(n -> {
            n.with("xt/document-store", docStore -> {
                docStore.module("xtdb.s3/->document-store");
                docStore.set("bucket", "my-bucket");
                docStore.set("prefix", "my-prefix");
            });
        });
        // end::override-module[]

        close(xtdbNode);
    }

    @Test
    public void nestedModules() {
        // tag::nested-modules-0[]
        IXtdb xtdbNode = IXtdb.startNode(n -> {
            n.with("xt/tx-log", txLog -> {
                txLog.with("kv-store", kv -> {
                    kv.module("xtdb.rocksdb/->kv-store");
                    kv.set("db-dir", new File("/tmp/rocksdb"));
                });
            });
            // end::nested-modules-0[]
            /*
            This obviously won't compile so has to be commented out
            // tag::nested-modules-1[]
            n.with("xt/document-store", docStore -> { ... });
            n.with("xt/index-store", indexStore -> { ... });
            // end::nested-modules-1[]
             */
            // tag::nested-modules-2[]
        });
        // end::nested-modules-2[]

        close(xtdbNode);
    }

    @Test
    public void sharingModules() {
        // tag::sharing-modules[]
        IXtdb xtdbNode = IXtdb.startNode(n -> {
            n.with("my-rocksdb", rocks -> {
                rocks.module("xtdb.rocksdb/->kv-store");
                rocks.set("db-dir", new File("/tmp/rocksdb"));
            });
            n.with("xt/document-store", docStore -> {
                docStore.with("kv-store", "my-rocksdb");
            });
            n.with("xt/tx-log", txLog -> {
                txLog.with("kv-store", "my-rocksdb");
            });
        });
        // end::sharing-modules[]

        close(xtdbNode);
    }

    private void close(IXtdb xtdbNode) {
        try {
            xtdbNode.close();
        }
        catch (IOException e) {
            fail();
        }
    }
}
